### PR TITLE
new: `nim --eval:cmd` to run a command directly; also fixes #15731

### LINKED ISSUE #15731: `echo 'echo 1/3' | nim e -` fails
`echo 'echo 1/3' | nim e -` fails

### Example
`echo 'echo 1/3' | nim e -`

### Current Output
Error: not a NimScript file: stdinfile

### Expected Output
works

### Additional Information
devel 1.5.1 86d7726fb8044d4e2fc55b0eac3d1d42f5ffc94f

### PR DESCRIPTION
**Pull request by timotheecour**
## why?
* so I don't have to reach for python/D/node/bash next time I want to run a quick one-off one-liner, or open a nim file just for something I'll use once, and because I can write this faster in nim than looking up bash specific commands:
```
# using c backend:
find . | nim -e:'import os,times; for a in stdin.lines: (if a.getLastAccessTime > getTime()-1.hours: echo a)'

# or as a fast calculator thanks to nimscript:
nim e -e:'echo 1/3'
```

## common in other languages
* python: `python3 -c 'print(1+1)'`
* D: `rdmd --eval 'writeln(1+1)'`
* node: `node -e 'console.log(1+1)'`
* mongo: `mongo --eval '1+1'`
* ruby: `ruby -e 'print(1+1)'`
* perl: `perl -e 'print 1+1'`
* bash: `bash -c 'echo 1'`
* matlab, octave: `matlab -r 'disp(1+1)'`
etc... (even if those also have a `pipe code from stdin` variant )

## often better than the existing `echo cmd | nim r -`
* can be used with a shell command
* `echo cmd | nim r -` **prevents using stdin**, which prevents many useful cases (streaming based) in one-off commands. This new feature allows using stdin, eg:
```
`yes | nim -e:'for a in stdin.lines: echo (a,)'`
find . | nim -e:'import os; for a in stdin.lines: echo a.splitFile.ext'
```

* `echo cmd | nim r -` already exists but this new feature is simpler to use: less typing is essential for quick evaluation, eg:
```
echo "echo 1+1" | nim r -
vs
nim -e:"echo 1+1"
```
* furthermore, that's easy to edit on cmdline (press arrow up to recall previous cmd, then edit the cmd which is conveniently in last position)
* I'm now more inclined to use `nim` rather than python or D everytime I need a quick expression evaluator/calculator etc

## not redundant with `inim`
inim is very useful in general but doesn't replace the simplicity of this standalone, dependency free feature. Furthermore, inim can't be used like that, only as a pipe which has same drawbacks as `echo 'echo 1+1' | nim r -' (prevents using stdin etc)

## usage: works with all existing features (ie orthogonal design)
(or at least should, all combinations I tried work)

```bash
nim -e:cmd # implies `r` command, writes to $HOME/.cache/nim/cmdfile_d/cmdfile
nim -e:cmd - # - takes placeholder of the file position; same behavior as above
nim r -e:cmd - # `r` is now explicit; same behavior as above
nim js -r -e:cmd # this also works
nim r -b:js -e:cmd # ditto
nim doc -r -b:js -e:cmd -d:foo # ditto

# when a filename is given (instead of - or no positional argument), it's used to read in config nims/cfg files as usual
# this is consistent with other commands, eg `nim dump pathto/main.nim`
nim -e:cmd pathto/main.nim # will pretend pathto/main.nim contains `cmd` and read any config nims/cfg found as usual

# runtime arguments work too, but obviously consumes a file placeholder
nim -e:cmd - arg1 arg2 # commandLineParams().len would be 2
nim -e:cmd pathto/main.nim arg1 arg2 # explicit filename given
nim r -e:cmd pathto/main.nim arg1 arg2 # same as above

# also works in combination with nimscript, and is quite useful for the speed it offers
nim e -e:cmd
nim e -e:cmd -
nim e -e:cmd patho/main.nims # analog to `nim r -e:cmd patho/main.nim`, will pretend pathto/main.nims contains `cmd` and read any config nims/cfg found as usual
```

## newlines are also possible if it doesn't fit in a 1-liner
```bash
# this:
nim -e:'
echo 1
echo 2'

# or this (among other possibilities):
nim -e:"$(printf 'echo 1\necho 2')"
```

## honors `betterRun` to avoid recompiling when -e:cmd doesn't change
```
nim -e:cmd - foo
nim -e:cmd - bar # re-runs with a different runtime argument but doesn't recompile
```

## bikeshedding
I initially used `-i` (for `input`) but changed it to `-e` (for eval), which is consistent with majority of languages

## preexisting bugfix
`echo 'echo 1/3' | nim e -` now works; fixes https://github.com/nim-lang/Nim/issues/15731

### CODE REVIEW

File: changelog.md [hunk 1]
```diff
@@ -14,6 +14,8 @@
 
 - Added `--declaredlocs` to show symbol declaration location in messages.
 - Source+Edit links now appear on top of every docgen'd page when `nim doc --git.url:url ...` is given.
+add `--declaredlocs` to show symbol declaration location in messages
```
Comment (alaviss) [inline changelog.md]: Pretty sure this belong to an another PR.

File: compiler/commands.nim [hunk 2]
```diff
@@ -383,19 +383,32 @@ proc dynlibOverride(conf: ConfigRef; switch, arg: string, pass: TCmdLinePass, in
...
+proc handleCmdInput*(conf: ConfigRef) =
+  conf.projectName = "cmdfile"
+  handleStdinOrCmdInput()
+
 proc processSwitch*(switch, arg: string, pass: TCmdLinePass, info: TLineInfo;
                     conf: ConfigRef) =
   var
     key, val: string
   case switch.normalize
+  of "e": # switch `-e` (not to be confused with command `e` for nimscript)
```
Comment (alaviss) [inline compiler/commands.nim]: Can we have a long form of `eval` for it?

**Review by alaviss — COMMENTED** [review]
Not sure how I feel about `-e` compiling and executing instead of using NimScript though.

File: changelog.md (hunk 1 above)
Comment (timotheecour) [reply changelog.md]: fixed

File: compiler/commands.nim (hunk 2 above)
Comment (timotheecour) [reply compiler/commands.nim]: done; I've now replaced the existing undocumented --eval which was a noop

**Comment by timotheecour** [comment]
PTAL

> Not sure how I feel about -e compiling and executing instead of using NimScript though.

I was initially hesitant about that but after more thought this is absolutely the right thing to do so I've changed default command from `r` to `e` (`nim r -e:'for a in stdin.lines: echo a'` is still available if you need compile+run).
rationale:
* interactive speed is more important here as a default; it's now a usable calculator among other use cases
* c backend (etc) is still available, eg `nim r -e:'for a in stdin.lines: echo a'` or any other command
* a very minor reason is e is close to -e

File: compiler/nimconf.nim [hunk 3]
```diff
@@ -248,11 +248,16 @@ proc loadConfigs*(cfg: RelativeFile; cache: IdentCache; conf: ConfigRef; idgen:
     if readConfigFile(configPath, cache, conf):
       configFiles.add(configPath)
 
-  template runNimScriptIfExists(path: AbsoluteFile) =
+  template runNimScriptIfExists(path: AbsoluteFile, isMain = false) =
     let p = path # eval once
-    if fileExists(p):
+    var s: PLLStream
+    if isMain and conf.command == "e":
```
Comment (Araq) [inline compiler/nimconf.nim]: Wait what? This has no place in ``nimconf.nim``.

File: compiler/vm.nim [hunk 4]
```diff
@@ -2110,11 +2110,14 @@ proc evalStmt*(c: PCtx, n: PNode) =
   if c.code[start].opcode != opcEof:
     discard execute(c, start)
 
-proc evalExpr*(c: PCtx, n: PNode): PNode =
-  let n = transformExpr(c.graph, c.idgen, c.module, n)
-  let start = genExpr(c, n)
-  assert c.code[start].opcode != opcEof
-  result = execute(c, start)
+when false:
```
Comment (Araq) [inline compiler/vm.nim]: Nitpicking but this has nothing to do with the PR.

File: doc/basicopt.txt [hunk 5]
```diff
@@ -31,6 +31,9 @@ Options:
   --app:console|gui|lib|staticlib
                             generate a console app|GUI app|DLL|static library
   -r, --run                 run the compiled program with given arguments
+  -e, --eval:cmd            evaluates nim code directly; e.g.: `nim -e:"echo 1"`
```
Comment (Araq) [inline doc/basicopt.txt]: Please only introduce the long version, I cannot remember all these single char abbreviations anymore.

**Comment by Araq** [comment]
A new tool like ``nimeval`` that covers this and also "nim secret" would be an alternative solution. This tool can then also use linenoise or other dependencies more freely.

File: compiler/vm.nim (hunk 4 above)
Comment (timotheecour) [reply compiler/vm.nim]: I've now un-disabled it but it did relate to this PR in the following weak sense: `--eval` used to (IIRC) call `evalExpr` some time ago and `--eval` became undocumented, but that link got cut at some point so `evalExpr` became deadcode; now that this PR takes over `--eval` with a new meaning, it made sense to seal the fate of this unused deadcode

File: doc/basicopt.txt (hunk 5 above)
Comment (timotheecour) [reply doc/basicopt.txt]: done

File: compiler/nimconf.nim (hunk 3 above)
Comment (timotheecour) [reply compiler/nimconf.nim]: `conf.command == "e"` is to fix https://github.com/nim-lang/Nim/issues/15731 (+ an analog error for the new `--eval`), and the `isMain` part is to avoid evaluating `cmdInput` for each config.nims processed.

**Comment by timotheecour** [comment]
PTAL all comments addressed

> A new tool like nimeval that covers this and also "nim secret" would be an alternative solution. This tool can then also use linenoise or other dependencies more freely.

a separate `nimeval` could be later developped and co-exist with this and add more bells and whistles that don't belong in `nim`; but this is a useful standalone (only requires `nim`) that in almost all cases is a better alternative to the existing `echo cmd | nim r -` (or, after bugfix in this PR, `echo cmd | nim e -`). 

`nimeval` can be discussed separately (eg, using compiler as a library vs shelling out to a nim specified as `--nim`, etc, both of which have pros and cons)

**Comment by Varriount** [comment]
Why not just have an "eval" subcommand?

**Comment by timotheecour** [comment]
> Why not just have an "eval" subcommand?

for separation of concerns and compositionality. `--eval:cmd` is just responsible for defining the stream to run (as opposed to the standard fileStream or the stdinStream from `-`), but is otherwise designed to be **usable with all other commands**, eg:
```sh
nim e --eval:'echo 1+1' # use nimscript (e is implied so `nim e --eval:'echo 1+1'` works too as shortcut)
nim r --eval:'for a in stdin.lines: echo a' # uses `r` command; very useful since nimscript doesn't work with many features including this command
nim js -r --eval:import jsconsole; ...' # run some js
nim doc -r --eval:'include os' # who knows
```

**Comment by Araq** [comment]
> a separate nimeval could be later developped 

But then this "later" never happens.

**Comment by timotheecour** [comment]
> But then this "later" never happens.

I don't see why those 2 things are mutually exclusive.

`nim --eval` is meant as a replacement for the existing stdin input `echo cmd | nim r -` (for reasons already given in OP, including ability to use stdin inside eval cmd), at least most of the time you'd prefer `nim --eval` than `echo cmd | nim r -`; 

it runs one-off commands and doesn't add any dependencies, and can use any command (most useful being `r` (codegen) and `e` (VM)).

`nimeval` is something much more involved as it needs to handle, at the minimum:
* a session
* integrate with (eg) linenoise
* history
and maybe down the road, ability to un-import, remove variables etc;

that would be a replacement for `nim --secret`, not `nim --eval`, at least if it's meant to run on the VM. 

And then there's `inim` which was developped (despite `nim --secret` and `echo cmd | nim r -` existing), and is running via codegen instead of VM, but has its own issues (not a true repl since the whole program is compiled+re-run at each new cmd) but is nevertheless useful in the meantime.

I've also done work on a repl that uses codegen but doesn't compile+run the whole program, but that's not ready for release (I've in the past done this for D, using an approach that compiles each new input command and dlopen's it inside a single running instance, ie different from how inim works).

File: compiler/nimconf.nim (hunk 3 above)
Comment (Araq) [reply compiler/nimconf.nim]: Bugfix or not, it's not nimconf's business to do special logic for ``conf.command == "e"``, this should be done in an upper layer.

Comment (timotheecour) [reply compiler/nimconf.nim]: fixed

**Comment by timotheecour** [comment]
@araq PTAL I just fixed remaining comment

File: compiler/nimconf.nim [hunk 6]
```diff
@@ -248,11 +248,16 @@ proc loadConfigs*(cfg: RelativeFile; cache: IdentCache; conf: ConfigRef; idgen:
...
-  template runNimScriptIfExists(path: AbsoluteFile) =
+  template runNimScriptIfExists(path: AbsoluteFile, isMain = false) =
     let p = path # eval once
-    if fileExists(p):
+    var s: PLLStream
+    if isMain and optWasNimscript in conf.globalOptions:
+      if conf.projectIsStdin: s = stdin.llStreamOpen
+      elif conf.projectIsCmd: s = llStreamOpen(conf.cmdInput)
+    if s == nil and fileExists(p): s = llStreamOpen(p, fmRead)
+    if s!=nil:
```
Comment (Araq) [inline compiler/nimconf.nim]:
```suggestion
    if s != nil:
```
